### PR TITLE
[WIP] TypeScript support in REPL

### PR DESCRIPTION
**Pull request by bartlomieju**
Ref #1158 

This is initial pass at implementing TypeScript support in REPL. It's quite complex task and we're aiming for MVP with this PR.

There are a few PR that might need to land before this one including:
#1136 #1165 

cc: @kitsonk @hayd @ry

### CODE REVIEW

File: js/compiler.ts [hunk 1]
```diff
@@ -469,6 +469,31 @@ export class DenoCompiler
...
+    sourceCode: SourceCode,
+    previousOutput?: OutputCode
+  ): { diagnostics: ts.Diagnostic[], outputCode: OutputCode, additionalCode: OutputCode } {
+        
+    console.log('incrementalCompile', sourceCode, previousOutput);
+      
+    // const service = this._service;
+
+    // TODO:
+    // 1. transpile code to JS
```
Comment (bartlomieju) [inline js/compiler.ts]: My idea was to start here with `const output = ts.transpileModule(sourceCode, {});` but I'm having trouble running that

**Comment by bartlomieju** [comment]
@kitsonk PTAL any feedback would be much appreciated

**Comment by bartlomieju** [comment]
Right now I'm getting this diagnostic:
```
Option '--resolveJsonModule' cannot be specified without 'node' module resolution strategy.
```

Also since I used the same compiler options as for standard program execution which uses AMD, `define` is required to run the code (running any code results in: `ReferenceError: define is not defined`)

**Comment by hayd** [comment]
I wonder if the meat of this should move to vm.ts similar to node. https://nodejs.org/api/vm.html

**Comment by bartlomieju** [comment]
I've got basic compilation working, now is time for imports 👍

File: js/compiler.ts [hunk 2]
```diff
@@ -405,6 +405,7 @@ export class DenoCompiler
       assert(
         mediaType === MediaType.TypeScript || mediaType === MediaType.JavaScript
       );
+      console.log("pre emit output");
```
Comment (hayd) [inline js/compiler.ts]: before shipping, these should be util's log (hidden behind debug flag).

Comment (bartlomieju) [reply js/compiler.ts]: Of course, just needed some guidance during debugging.

**Comment by bartlomieju** [comment]
@hayd yes, that's what we're aiming for. Deno compiler will return compiled JS code and it will be up to REPL to execute it.

**Comment by hayd** [comment]
Is this waiting on #1136?

**Comment by kitsonk** [comment]
Actually I think we need to wait for #1190 which I am working on at the moment, but it is fairly complex.  Once we have that, we can create a better incremental compiler too as well.

**Comment by ry** [comment]
Closing old PRs. I'm still not ready to address TS in the REPL - but now that native ES modules are landed we're much closer. Let's talk about it in a few weeks.

**Pull request closed without merging**